### ISSUE TITLE
How to handle DSL codebases that inject names via from unresolved_pkg import *?

### ISSUE THREAD

**Issue by mathomp4** [astral-sh/ty#3566]
## Note

I freely admit Claude helped me write this. I'm trying to learn Python typing, but, well, explaining what I was seeing is way above my head. I mean, I'm just trying not to use `Any` everywhere. :)

---

## Summary

Some Python codebases use `from some.package import *` as a deliberate DSL mechanism that injects names into every module's scope. The example I'm dealing with are [Spack](https://spack.io/) package repositories, where every `package.py` starts with:

```python
from spack.package import *
```

This injects hundreds of names (`license`, `version`, `depends_on`, `variant`, etc.) that form the entire authoring API. When `spack` is not installed in the type-checking environment, ty cannot resolve this import, and names fall through to whatever ty finds elsewhere — including Python builtins.

## The problem

When `spack` is unresolvable, `license` (a spack directive) falls through to Python's builtin `license` object (a `_Printer`). ty resolves it as that builtin and produces diagnostics that are technically correct for `_Printer.__call__`, but not what was intended:

```
error[too-many-positional-arguments]: Too many positional arguments to bound method `_Printer.__call__`: expected 1, got 2
  --> package.py:18:13
   |
18 |     license("Apache-2.0", checked_by="mathomp4")

error[unknown-argument]: Argument `checked_by` does not match any known parameter of bound method `_Printer.__call__`
```

The root cause is that names from an unresolvable `import *` fall through to ambient scope (including builtins) rather than being treated as `Unknown`.

## Current workaround

**`__builtins__.pyi`** (via [ruff#22021](https://github.com/astral-sh/ruff/pull/22021)) — declare the clashing name as `Any`:

```python
# __builtins__.pyi
from typing import Any
license: Any
```

This works, but requires manually identifying which injected names clash with Python builtins and re-declaring them globally as `Any`. In this case only `license` clashes, but for a larger or different spack repo there could be more, and there's no easy way to know without running `ty check` and hunting down each false positive.

You can see the my full `ty`/`ruff` config at: https://github.com/GMAO-SI-Team/geosesm-spack

## Is there a better way?

Is there a cleaner intended way to handle this pattern? Some ideas Claude had because this is *deep* Python now:

1. **`replace-imports-with-any`** for wildcard imports — currently `replace-imports-with-any` suppresses the import diagnostic but names still fall through to builtins. If names from a `replace-imports-with-any` module were treated as `Unknown`/`Any` rather than falling through, this would be solved cleanly.
2. A **`builtins` config option** (analogous to ruff's [`builtins` setting](https://docs.astral.sh/ruff/settings/#builtins)) — a list of names to treat as `Any` in the builtins namespace, without needing a `.pyi` file.
3. Names from an **unresolvable `import *`** being treated as `Unknown` rather than falling through to ambient scope.

## Related

- #1297 — similar context (IPython builtins in notebooks injected without an import)
- [ruff#22021](https://github.com/astral-sh/ruff/pull/22021) — adds `__builtins__.pyi` support (current workaround)
- [spack/spack-packages](https://github.com/spack/spack-packages) — the upstream project using this pattern at scale
- [spack/spack#52446](https://github.com/spack/spack/pull/52446) — WIP PR by the spack team to add ty to `spack style` (replacing mypy) in the core spack codebase

### Version

ty 0.0.39

**Comment by AlexWaygood** [comment]
It sounds like the fundamental problem here is that ty cannot resolve your `from spack.package import *` import! ty supports `*` imports, so, unless `Spack` is using some very dynamic mechanism to expose all these symbols, we should be able to understand that all these symbols are present in your module.

What's the specific diagnostic you're seeing on that line from ty? Have you tried to debug why ty can't seem to resolve this import? Do you definitely have it installed in the virtual environment ty is resolving for your project?

**Comment by mathomp4** [comment]
@AlexWaygood So the error I get is:

```
❯ ty check
error[too-many-positional-arguments]: Too many positional arguments to bound method `_Printer.__call__`: expected 1, got 2
  --> spack_repo/geosesm/packages/geosgcm_deps/package.py:18:13
   |
18 |     license("Apache-2.0", checked_by="mathomp4")
   |             ^^^^^^^^^^^^
   |
info: Method signature here
  --> stdlib/_sitebuiltins.pyi:22:9
   |
22 |     def __call__(self) -> None: ...
   |         ^^^^^^^^^^^^^^^^^^^^^^
   |

error[unknown-argument]: Argument `checked_by` does not match any known parameter of bound method `_Printer.__call__`
  --> spack_repo/geosesm/packages/geosgcm_deps/package.py:18:27
   |
18 |     license("Apache-2.0", checked_by="mathomp4")
   |                           ^^^^^^^^^^^^^^^^^^^^^
   |
info: Method signature here
  --> stdlib/_sitebuiltins.pyi:22:9
   |
22 |     def __call__(self) -> None: ...
   |         ^^^^^^^^^^^^^^^^^^^^^^
   |

Found 2 diagnostics
```

NOTE: this is already with:
```toml
[tool.ty.src]
include = [
    "spack_repo/geosesm/packages/*/package.py",
]

[tool.ty.analysis]
# Spack packages use "from spack.package import *" which injects many names into scope.
# ty cannot resolve these statically, so suppress unresolved-import diagnostics for
# spack and llnl modules.
allowed-unresolved-imports = [
    "spack.**",
    "llnl.**",
]

[tool.ty.rules]
# Mirrors mypy's disable_error_code = ["no-redef", "name-defined", "attr-defined"]
# "no-redef"      -> conflicting-declarations (spack packages redefine names via import *)
# "name-defined"  -> unresolved-reference, possibly-unresolved-reference
# "attr-defined"  -> unresolved-attribute, possibly-missing-attribute
conflicting-declarations = "ignore"
unresolved-reference = "ignore"
possibly-unresolved-reference = "ignore"
unresolved-attribute = "ignore"
possibly-missing-attribute = "ignore"
# Also ignore missing imports globally (spack itself may not be on the path)
unresolved-import = "ignore"
```

The issue is [my repo](https://github.com/GMAO-SI-Team/geosesm-spack) is sort of a "user" version of [spack-packages](https://github.com/spack/spack-packages/) where I have packages that really no one in the spack community would care about (for now?).

Now, when you run `spack command` it handles the fact that my repo is outside and all, but it works because it's running python, etc.

But `ty` has no idea that say on this machine I have my spack repo in `~/spack-repos/spack` say. And it might be different on different machines. 

I hate to do it, but someone like @johnwparent might know more. He's the one doing the ty-ification of spack (see https://github.com/spack/spack/pull/52446) and maybe in the end, his other PR for `spack style` (https://github.com/spack/spack/pull/52161) will make it moot.

But at least with my changes/hacks for `ty` in my repo, I can do `ty check` alone. :)

**Comment by carljm** [comment]
Thanks for the report!

I don't think it's really possible to support this case in the way that you want. Fundamentally we have to know what names are imported by `from spack import *`, and if we can't find the `spack` module, there is no way we can possibly figure that out. We can't just assume `from spack import *` imports _all_ names, or else suddenly you'd find that `int` and `str` and `list` and `dict` and `foobar_does_not_exist_anywhere` and a million other things that are _not_ ever actually defined by spack are also resolving as `Unknown` without error. Your option 3 mentions special handling for "Names from an unresolvable `import *` ..." -- but the whole problem here is that if the import is not resolvable, we have no idea which names those might be.

It's not clear to me why you need or want to type-check your spack-dependent project without spack available to the type checker. If you need to use a custom local spack repo, the usual way to handle this would be to editable-install that repo into an environment which ty can see.

If you really can't make spack visible to the type checker, then I think the stub file idea is the right direction, but it shouldn't be a builtins stub file, it should just be a `spack.pyi` stub file, which ty would happily consider as the source of truth for the contents of the `spack` module. That would also allow you to define better types than just `Any` if you want, and get more value from the type check.

Going to close this as "not planned", since I don't see an improvement we can make in ty here. (But feel free to suggest something if you do see one.)

**Comment by johnwparent** [comment]
@mathomp4 Just to close the loop from the Spack end here. 

For the packages repo + `ty`, a couple extra things will come into play here, running `spack style -t ty` will just work for the packages repo (once I plumb it up) because while the packages repo doesn't know where Spack is, spack itself does, so we can provide the resolution for those imports. Further, we'll be using a typing stub file for the package API (everything packages will need to import via `from spack import *` ) that `ty` will have access to if you're just running ty on the repos without Spack. 

We'll be including that tooling to generate or copy that typing stub for your own package repos (or just run `spack style -t --repo`).

Edit: I agree with the ty folks, this is a repo linting setup issue, not so much a ty issue.

Also since I'm here, thanks for ty, ty folks!
